### ISSUE TITLE
How can I avoid 拖

### ISSUE THREAD

**Issue by Boreaso** [mockito/mockito#2398]
> Hey,
>
> First thanks for reporting, in order to help us to classify issue can you make sure the following check boxes are checked ?
>
> If this is about mockito usage, the better way is to reach out to
>
>  - stackoverflow : https://stackoverflow.com/questions/tagged/mockito
>  - the mailing-list  : https://groups.google.com/forum/#!forum/mockito / mockito@googlegroups.com
>    (Note mailing-list is moderated to avoid spam)
>
> _This block can be removed_
> _Something wrong in the template fix it here `.github/ISSUE_TEMPLATE.md`

check that

 - [ ] The mockito message in the stacktrace have useful information, but it didn't help
 - [ ] The problematic code (if that's possible) is copied here;
       Note that some configuration are impossible to mock via Mockito
 - [ ] Provide versions (mockito / jdk / os / any other relevant information)
 - [ ] Provide a [Short, Self Contained, Correct (Compilable), Example](http://sscce.org) of the issue
       (same as any question on stackoverflow.com)
 - [ ] Read the [contributing guide](https://github.com/mockito/mockito/blob/main/.github/CONTRIBUTING.md)

**Comment by wonsuc** [comment]
Any plan to implement this?

**Comment by raphw** [comment]
No, not really. This would cause that a class would only work in "mock mode". This is not what this feature aims to achieve.

**Comment by Delfic** [comment]
How can I leverage `Mockito.mockStatic`?
It runs the real methods called in any `static` before it mocks them, since calling the `mockStatic` method initializes the class.

**Comment by raphw** [comment]
It's not possible. To mock a class, we need to load it. Also unmocking would alter the class forever what goes against the overall philosophy.

**Comment by hitime888** [comment]
Do you have any plans to support this?I have some scenarios that require disabling static initialization.For example, static initialization is very time-consuming, and I want to start unit testing as soon as possible.

**Comment by raphw** [comment]
No, this would affect global state and would be indeterministic as Mockito does not control class loading.

**Comment by mirkoalicastro** [comment]
Hi @raphw , it's me again 🙂. I have been working on this feature since Mockito 5 is now supposed to be loaded as a java agent, it is now possible to suppress the `<clinit>`. Any chance you could have a look at the idea and implementation? I've detailed the PR description: https://github.com/mockito/mockito/pull/3798. Thank you!

**Comment by raphw** [comment]
The issue I have with this approach is that it depends on class loading
order. I would want this to work no matter how.

I think this would need to work as follows:

1. Mockito is loaded as a Java agent.
2. The Java agent comes with an argument that mentions that static
initializers should be mocked.
3. The agent immediately adds a class file transformer.
4. The class file transformer mocks the static initializers of all classes
that are annotated with some (at)MockitoInitializerSuppression

This will then create stable results, but I do not think this should be
added to the regular API.

Would you be interested in attempting this? I am happy to guide you through
this.

**Comment by mirkoalicastro** [comment]
Hi @raphw, thank you for your feedback.

With the current approach, using forkEvery = 1 (Gradle) / reuseForks=false (Maven) gives each test class its own JVM, which eliminates cross-class contamination and leads to stable results:
  - Test A suppresses Foo -> won't leak to test B (different JVM)
  - Test B loading Foo normally first -> won't block test A from suppressing it (different JVM)

Tests could also be grouped if build time is a concern. Do you think this approach is still worth considering, or should it be disregarded entirely?

---

That said, your point still holds if the forkEvery = 1 (Gradle) / reuseForks=false (Maven) should not be considered at all. If I understand correctly, your suggestion would involve configuring suppression via agent arguments so that the transformer is active from the very start of the JVM and for all the tests. This would guarantee that the specified classes always have their static initializers suppressed, regardless of when or how they get loaded. That would still be very useful when dealing with legacy code that brings heavy static initialization that developers never want to run during unit tests.

If you're in favour of this second approach, I would be happy to give it a try with some guidance.

**Comment by mirkoalicastro** [comment]
Hey @raphw . 

I've worked towards your idea and solution is actually much simpler and it produces stable results - suppressing the clinit for all the tests. I've raised a PR for it (https://github.com/mockito/mockito/pull/3801) . Any chance you could have a look at it? 🙏

**Comment by hmadhkhwjly71-cell** [comment]
هادي 2###
